base/reduce.jl

```diff
@@ -362,10 +367,95 @@ julia> reduce(*, [2; 3; 4]; init=-1)
 -24
 ```
 """
-reduce(op, itr; kw...) = mapreduce(identity, op, itr; kw...)
+function reduce(op, itr::T; kw...) where T
+    # Redispatch, adding traits
+    reduce(op, itr, eltype_or_default_eltype(itr), IteratorSize(T); kw...)
```

Comment:
This should be a private `_reduce` method.

Reply:
Yes, unless it shoubd be a private `_mapreduce` methods.
What do you think?

Reply:
I guess `_mapreduce` is better if you can support that, since that will also support `mapreduce(identity, ...)`.